### PR TITLE
[cli] Added `EXPERIMENTAL_VITE` flag to start `devServer` with `vite` instead of `webpack`

### PR DESCRIPTION
**Pull request by benjdlambert**
This is in the experimentation phase, but we've discovered internally at Spotify that the performance of webpack for development purposes is getting to a point where it's become really slow for both hot reloads and startup.

We experimented internally at switching to Vite, and the performance is much better and hot reloading is instant using react-refresh.

This is still under heavy development, and I think there's still some improvements we can make to the vite config here to make it even faster.

### CODE REVIEW

File: packages/cli/src/lib/bundler/server.ts [hunk 1]
```diff
@@ -139,58 +161,99 @@ export async function serveBundle(options: ServeOptions) {
...
+    } else {
+      (server as WebpackDevServer).startCallback((err?: Error) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+      });
+    }
+    openBrowser(url.href);
+    resolve();
```
Comment (vinzscam) [inline packages/cli/src/lib/bundler/server.ts:256]: just noticed that the `resolve` function might be called twice at this point

Comment (benjdlambert) [reply packages/cli/src/lib/bundler/server.ts:257]: nice catch

File: packages/cli/src/lib/bundler/server.ts [hunk 2]
```diff
@@ -139,58 +161,99 @@ export async function serveBundle(options: ServeOptions) {
...
-      historyApiFallback: {
-        // Paths with dots should still use the history fallback.
-        // See https://github.com/facebookincubator/create-react-app/issues/387.
-        disableDotRule: true,
-
-        // The index needs to be rewritten relative to the new public path, including subroutes.
-        index: `${config.output?.publicPath}index.html`,
+      plugins: [
+        react(),
+        vitePluginSvgr(),
```
Comment (Rugvip) [inline packages/cli/src/lib/bundler/server.ts]: I think we could skip this for now if it causes trouble tbh, it's just for the `.icon.svg` support right? Or is it to be able to import `.svg` as an external resource ref too?

**Review by Rugvip — APPROVED** [review]
Nice 😎 

And very slim 😁

File: packages/cli/src/lib/bundler/server.ts (hunk 2 above)
Comment (Rugvip) [reply packages/cli/src/lib/bundler/server.ts]: Ah, outdated one from earlier, I assume this is what you did? :p

File: packages/cli/src/lib/bundler/server.ts [hunk 3]
```diff
@@ -139,58 +160,102 @@ export async function serveBundle(options: ServeOptions) {
...
 
-  const compiler = webpack(config);
-
-  server = new WebpackDevServer(
-    {
-      hot: !process.env.CI,
-      devMiddleware: {
-        publicPath: config.output?.publicPath as string,
-        stats: 'errors-warnings',
+  console.log(paths.targetHtml);
```
Comment (freben) [inline packages/cli/src/lib/bundler/server.ts:163]: remove

File: packages/cli/src/lib/bundler/server.ts [hunk 4]
```diff
@@ -139,58 +160,102 @@ export async function serveBundle(options: ServeOptions) {
...
+  if (process.env.EXPERIMENTAL_VITE) {
+    server = await vite.createServer({
+      define: {
+        global: 'globalThis',
```
Comment (freben) [inline packages/cli/src/lib/bundler/server.ts:167]: literally that string? i presume it's an enum so vite does something sensible internally

File: packages/cli/src/lib/bundler/server.ts [hunk 5]
```diff
@@ -139,58 +160,102 @@ export async function serveBundle(options: ServeOptions) {
...
+        'process.argv': JSON.stringify(process.argv),
+        'process.env.APP_CONFIG': JSON.stringify(cliConfig.frontendAppConfigs),
+      },
+      resolve: {
+        alias: {
+          'node-fetch': 'cross-fetch',
```
Comment (freben) [inline packages/cli/src/lib/bundler/server.ts:173]: Maybe worth a comment about why this is here

**Review by freben — COMMENTED** [review]
Alrigt!

File: packages/cli/src/lib/bundler/server.ts (hunk 2 above)
Comment (benjdlambert) [reply packages/cli/src/lib/bundler/server.ts]: ye

File: packages/cli/src/lib/bundler/server.ts (hunk 4 above)
Comment (benjdlambert) [reply packages/cli/src/lib/bundler/server.ts:167]: yeah `globalThis === window` in the browser. but it's also defined in node land too. Gonna try and change this to `window` to see if it still works to clear maybe some confusion.

File: packages/cli/src/lib/bundler/server.ts [hunk 6]
```diff
@@ -139,58 +160,99 @@ export async function serveBundle(options: ServeOptions) {
...
+      (server as WebpackDevServer).startCallback((err?: Error) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+        openBrowser(url.href);
+        resolve();
+      });
+    }
   });
```
Comment (acierto) [inline packages/cli/src/lib/bundler/server.ts:256]: Can be a bit optimized to:
```

await new Promise<void>(async (resolve, reject) => {
    if (process.env.EXPERIMENTAL_VITE) {
      await (server as vite.ViteDevServer).listen();
    } else {
      (server as WebpackDevServer).startCallback((err?: Error) => {
        if (err) {
          reject(err);
          return;
        }
      });
    }
    openBrowser(url.href);
    resolve();
  });
```

File: packages/cli/src/lib/bundler/server.ts [hunk 7]
```diff
@@ -123,7 +136,15 @@ export async function serveBundle(options: ServeOptions) {
...
-      server?.invalidate();
+      if (server) {
+        if ('invalidate' in server) {
+          server.invalidate();
+        }
+
+        if ('restart' in server) {
+          server.restart();
+        }
+      }
```
Comment (acierto) [inline packages/cli/src/lib/bundler/server.ts:147]: How about to create an extra function for this, as it is duplicated twice, same as on line `94-102`

File: packages/cli/package.json [hunk 8]
```diff
@@ -132,6 +133,9 @@
     "tar": "^6.1.12",
     "terser-webpack-plugin": "^5.1.3",
     "util": "^0.12.3",
+    "vite": "^4.4.9",
+    "vite-plugin-html": "^3.2.0",
+    "vite-plugin-node-polyfills": "^0.14.1",
```
Comment (Rugvip) [inline packages/cli/package.json:138]: Had a thought, perhaps we should make some of these peer deps while it's still in this phase? At least the large ones. `vite-plugin-html` and `@vitejs/plugin-react` are both slim, but `vite` is over 3 MB and brings in `esbuild`, and somehow `vite-plugin-node-polyfills` is over 13 MB. Apparently it's all examples shipped in the package itself, opened https://github.com/davidmyersdev/vite-plugin-node-polyfills/pull/50. Ideally that would get merged and published soonish and we could just leave `vite` as a peer dep?

File: packages/cli/src/lib/bundler/server.ts (hunk 7 above)
Comment (Rugvip) [reply packages/cli/src/lib/bundler/server.ts:147]: Could even separate the two:

```
webpackServer?.invalidate()
viteServer?.restart()
```

File: packages/cli/src/lib/bundler/server.ts (hunk 1 above)
Comment (Rugvip) [reply packages/cli/src/lib/bundler/server.ts:252]: More importantly this changes the logic for the webpack server, I think it's best to duplicate this in the two if blocks tbh

**Review by Rugvip — APPROVED** [review]
image

**Comment by benjdlambert** [comment]
:tada:

**Pull request merged**